### PR TITLE
Flexible date formats for Timelines

### PR DESCRIPTION
**Pull request by annebyrne**
Allow users to select relevant date format for events and date ranges if required.

Tested on `CODE`

screen shot 2017-11-15 at 11 42 59

### CODE REVIEW

File: public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js [hunk 1]
```diff
@@ -3,30 +3,74 @@ import {ManagedForm, ManagedField} from '../../../ManagedEditor';
...
+      body: PropTypes.string,
+      dateFormat: PropTypes.string
     }),
     fieldPlaceholder: PropTypes.string,
     onUpdateField: PropTypes.func,
     onFormErrorsUpdate: PropTypes.func
   };
 
+    state = {
+        dateRangeRequired: typeof this.props.fieldValue === PropTypes.shape({
```
Comment (tomrf1) [inline public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js]: what's happening here, is it checking for the existence of `toDate`?

Comment (annebyrne) [reply public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js]: Yes it is. On first render, `fieldValue` is undefined so this is why `this.props.fieldValue.toDate` doesn't work and we require a more verbose solution

**Comment by annebyrne** [comment]
There is a bug in React which means that multiple radio button lists will cancel the `checked` attribute of all buttons save one. Because of this, we've chosen to go with a select box for now until the problem is resolved.

It should hopefully be sorted soon so the relevant code will be uncommented and removed.

File: public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js (hunk 1 above)
Comment (regiskuckaertz) [reply public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js]: have you tried `this.props.fieldValue && 'toDate' in this.props.fieldValue`?

File: public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js [hunk 2]
```diff
@@ -3,30 +3,85 @@ import {ManagedForm, ManagedField} from '../../../ManagedEditor';
...
+       //  Remove ID in dateFormats array as will be superfluous
+       //
+       //  See this PR for further info: https://github.com/facebook/react/pull/11227
+       //
+       //  The radio button code has been left in for now as React devs are working on this bug.
+       //  React cannot handle multiple radio button lists and leaves only one button visibly checked
+       //  regardless of the props. This is only a bug of UX, and the data is being stored correctly
+       //  on the atom. We've chosen to release this feature with a select so to speed up delivery.
+
+dateFormats = [
```
Comment (regiskuckaertz) [inline public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js]: Best put these out of the component or at least as a static property.

File: public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js (hunk 1 above)
Comment (annebyrne) [reply public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js]: It works but produces a few warnings. I think I'm going to stick with the above for now.

File: public/js/components/FormFields/FormFieldRadioButtons.js [hunk 3]
```diff
@@ -1,40 +1,51 @@
...
   }
 
-  renderButton(value, i) {
+  renderButton(value, label, i) {
     return (
-      <div key={i} className="form__group--radio">
-        <input className="form__radio-btn" type="radio" name={this.props.fieldName} value={value} checked={this.props.fieldValue === value ? 'checked' : false} onChange={this.onUpdate} />
-        <span className="form__label form__label--radio">{value}</span>
+      <div key={i+uuidv4()} className="form__group--radio">
+        <input key={uuidv4()} className="form__radio-btn" type="radio" value={value} checked={this.props.fieldValue === value ? 'checked' : false} onChange={this.onUpdate} name={this.props.fieldName} />
```
Comment (tomrf1) [inline public/js/components/FormFields/FormFieldRadioButtons.js]: we can get rid of these uuids now that we've finished debugging

File: public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js [hunk 4]
```diff
@@ -35,6 +90,25 @@ export class TimelineItem extends React.Component {
           <ManagedField fieldLocation="date" name="Date" isRequired={true}>
             <FormFieldDateTextInput/>
           </ManagedField>
+          <FormFieldCheckbox fieldName="ranged-date" fieldLabel="Want a date range?" fieldValue={this.state.dateRangeRequired} onUpdateField={this.updateDateRange.bind(this)}/>
+          <ManagedField isRequired={false} fieldLocation="toDate">
```
Comment (regiskuckaertz) [inline public/js/components/AtomEdit/CustomEditors/TimelineFields/TimelineItem.js]: Another option would be to render this field only if `state.dateRangeRequired`

**Comment by regiskuckaertz** [comment]
Nice!

**Comment by tomrf1** [comment]
👍 🎉

**Pull request merged**